providers/implementations/ciphers/cipher_aes_gcm_siv_hw.c

```diff
@@ -0,0 +1,373 @@
+    }
+    polyval_ghash_hash(ctx->Htable, S_s, (uint8_t *)len_blk, TAG_SIZE);
+
+    for (i = 0; i < NONCE_SIZE; i++)
+        S_s[i] ^= ctx->nonce[i];
+
+    S_s[TAG_SIZE - 1] &= 0x7f;
+
+    /*
+     * In the ctx, user_tag is the one received/set by the user,
+     * and tag is generated from the input
+     */
+    out_len = sizeof(ctx->tag);
+    error |= !EVP_EncryptUpdate(ctx->ecb_ctx, ctx->tag, &out_len, S_s, sizeof(S_s));
+    ctx->generated_tag = !error;
+    /* Regardless of error */
+    ctx->used_dec = 1;
+    return !error;
+}
+
+static int aes_gcm_siv_cipher(void *vctx, unsigned char *out,
+                              const unsigned char *in, size_t len)
+{
+    PROV_AES_GCM_SIV_CTX *ctx = (PROV_AES_GCM_SIV_CTX *)vctx;
+
+    /* EncryptFinal or DecryptFinal */
+    if (in == NULL)
+        return aes_gcm_siv_finish(ctx);
+
+    /* Deal with associated data */
+    if (out == NULL)
+        return aes_gcm_siv_aad(ctx, in, len);
+
+    if (ctx->enc)
+        return aes_gcm_siv_encrypt(ctx, in, out, len);
+
+    return aes_gcm_siv_decrypt(ctx, in, out, len);
+}
+
+static void aes_gcm_siv_clean_ctx(void *vctx)
+{
+    PROV_AES_GCM_SIV_CTX *ctx = (PROV_AES_GCM_SIV_CTX *)vctx;
+
+    EVP_CIPHER_CTX_free(ctx->ecb_ctx);
+    ctx->ecb_ctx = NULL;
+}
+
+static int aes_gcm_siv_dup_ctx(void *vdst, void *vsrc)
+{
+    PROV_AES_GCM_SIV_CTX *dst = (PROV_AES_GCM_SIV_CTX *)vdst;
+    PROV_AES_GCM_SIV_CTX *src = (PROV_AES_GCM_SIV_CTX *)vsrc;
+
+    dst->ecb_ctx = NULL;
+    if (src->ecb_ctx != NULL) {
+        if ((dst->ecb_ctx = EVP_CIPHER_CTX_new()) == NULL)
+            goto err;
+        if (!EVP_CIPHER_CTX_copy(dst->ecb_ctx, src->ecb_ctx))
+            goto err;
+    }
+    return 1;
+
+ err:
+    EVP_CIPHER_CTX_free(dst->ecb_ctx);
+    dst->ecb_ctx = NULL;
+    return 0;
+}
+
+static const PROV_CIPHER_HW_AES_GCM_SIV aes_gcm_siv_hw =
+{
+    aes_gcm_siv_initkey,
+    aes_gcm_siv_cipher,
+    aes_gcm_siv_dup_ctx,
+    aes_gcm_siv_clean_ctx,
+};
+
+const PROV_CIPHER_HW_AES_GCM_SIV *ossl_prov_cipher_hw_aes_gcm_siv(size_t keybits)
+{
+    return &aes_gcm_siv_hw;
+}
+
+/* NEED AES-CTR32... which is a 32bit counter version of AES-CTR which is 128? */
+static int ossl_aes_gcm_siv_ctr32(PROV_AES_GCM_SIV_CTX *ctx, const unsigned char *init_counter,
+                                  unsigned char *out, const unsigned char *in, size_t len)
+{
+    uint8_t keystream[BLOCK_SIZE];
+    int out_len;
+    size_t i;
+    size_t j;
+    size_t todo;
+    uint32_t counter;
+    int error = 0;
+    union {
+        uint32_t x32[BLOCK_SIZE/sizeof(uint32_t)];
+        uint8_t x8[BLOCK_SIZE];
+    } block;
+    DECLARE_IS_ENDIAN;
+
+    memcpy(&block, init_counter, sizeof(block));
+    if (IS_BIG_ENDIAN) {
+        counter = BSWAP4(block.x32[0]);
+    }
+
+    for (i = 0; i < len; i += sizeof(block)) {
+        out_len = BLOCK_SIZE;
+        error |= !EVP_EncryptUpdate(ctx->ecb_ctx, keystream, &out_len, block.x8, sizeof(block));
```

Comment:
It looks like this breaks the strict aliasing rules.  Passing a pointer to a union field isn't allowed.

Yeah, crazy rule.

Reply:
Suggestions?

Reply:
Passing the union fixes things (I think, the standard isn't obvious).  The other alternative would be casting betwixt the two types.

Currently, the builds aren't strong on the strict aliasing front -- as compilers become stricter, we'll find breaks.

I will note that I think enforcement of things like this is crazy but it's happened already...